syft/source/source.go

```diff
@@ -97,34 +101,54 @@ func generateFileSource(fs afero.Fs, location string) (*Source, func(), error) {
 		return &Source{}, func() {}, fmt.Errorf("given path is not a directory (path=%q): %w", location, err)
 	}
 
-	s, err := NewFromFile(location)
+	s, cleanupFn, err := NewFromFile(location)
 	if err != nil {
 		return &Source{}, func() {}, fmt.Errorf("could not populate source from path=%q: %w", location, err)
 	}
 
-	return &s, func() {}, nil
+	return &s, cleanupFn, nil
 }
 
 // NewFromDirectory creates a new source object tailored to catalog a given filesystem directory recursively.
 func NewFromDirectory(path string) (Source, error) {
 	return Source{
-		Mutex: &sync.Mutex{},
+		mutex: &sync.Mutex{},
 		Metadata: Metadata{
 			Scheme: DirectoryScheme,
 			Path:   path,
 		},
+		path: path,
 	}, nil
 }
 
-// NewFromDirectory creates a new source object tailored to catalog a given filesystem directory recursively.
-func NewFromFile(path string) (Source, error) {
+// NewFromFile creates a new source object tailored to catalog a file.
+func NewFromFile(path string) (Source, func(), error) {
+	var analysisPath = path
+	var cleanupFn func()
+
+	// if the given file is an archive (as indicated by the file extension and not MIME type) then unarchive it and
+	// use the contents as the source. Note: this does NOT recursively unarchive contents, only the given path is
+	// unarchived.
+	envelopedUnarchiver, err := archiver.ByExtension(path)
+	if unarchiver, ok := envelopedUnarchiver.(archiver.Unarchiver); err == nil && ok {
+		unarchivedPath, tmpCleanup, err := unarchiveToTmp(path, unarchiver)
```

Comment:
I'm wondering if the cleanup function value should be captured as `cleanupFn` (already declared), rather than introduce a new variable here. `NewFromFile` returns a cleanup function, it could be useful to return the cleanup task definition that comes from `unarchiveToTmp` in the event of an error being returned from `unarchiveToTmp`. What are your thoughts here?

Reply:
We only want to replace `cleanupFn` with another function once we are certain that `unarchiveToTmp` has produced a function (there is no error returned).

Reply:
Thanks for the offline sync --I've updated this section to always ensure that a cleanup function will be provided.

Reply:
Looks good, thanks!